test/jdk/java/util/zip/ZipCoding.java

```diff
@@ -23,111 +23,173 @@
 
 /**
  * @test
- * @bug 4244499 4532049 4700978 4820807 4980042
+ * @bug 4244499 4532049 4700978 4820807 4980042 7009069 8322802
  * @summary Test ZipInputStream, ZipOutputStream and ZipFile with non-UTF8 encoding
  * @modules jdk.charsets
+ * @run junit ZipCoding
  */
 
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
 import java.io.*;
 import java.nio.charset.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.*;
+import java.util.stream.Stream;
 import java.util.zip.*;
 
+import static org.junit.jupiter.api.Assertions.*;
+
 public class ZipCoding {
 
-    public static void main(String[] args) throws Exception {
+    // The data to write to ZIP entries in this test
+    private static byte[] ENTRY_DATA = "German Umlaut \u00fc in entry data"
+            .getBytes(StandardCharsets.ISO_8859_1);
 
-        test("MS932",
-             "\u4e00\u4e01", "\uff67\uff68\uff69\uff6a\uff6b\uff6c");
+    /**
+     * Provide arguments used for parameterized tests
+     * @return a stream of argument lists
+     */
+    public static Stream<Arguments> charsetsAndNames() {
+        // Arguments are: Write charset, read charset, entry name, comment
+        return Stream.of(
+                // MS code page 932 for the Japanese language
+                Arguments.of("MS932", "MS932",
+                        "\u4e00\u4e01",
+                        "\uff67\uff68\uff69\uff6a\uff6b\uff6c"),
 
-        test("ibm437",
-             "\u00e4\u00fc", "German Umlaut \u00fc in comment");
+                // Code page for the IBM PC
+                Arguments.of("ibm437", "ibm437",
+                        "\u00e4\u00fc",
+                        "German Umlaut \u00fc in comment"),
 
-        test("utf-8",
-             "\u4e00\u4e01", "\uff67\uff68\uff69\uff6a\uff6b\uff6c");
+                // UTF-8 with Japanese characters
+                Arguments.of("utf-8", "utf-8",
+                        "\u4e00\u4e01",
+                        "\uff67\uff68\uff69\uff6a\uff6b\uff6c"),
 
-        test("utf-8",
-             "\u00e4\u00fc", "German Umlaut \u00fc in comment");
+                // UTF-8 with characters in the Latin1 range
+                Arguments.of("utf-8", "utf-8",
+                        "\u00e4\u00fc",
+                        "German Umlaut \u00fc in comment"),
 
-        test("utf-8",
-             "Surrogate\ud801\udc01", "Surrogates \ud800\udc00 in comment");
+                // UTF-8 with surrogate pairs
+                Arguments.of("utf-8", "utf-8",
+                        "Surrogate\ud801\udc01",
+                        "Surrogates \ud800\udc00 in comment"),
 
+                // ZipOutputStream sets the 'Language encoding flag' when writing using UTF-8
+                // UTF-8 should be used for decoding, even when opening with a different charset
+                Arguments.of("utf-8", "iso-8859-1",
```

Comment:
Hello Eirik, there are 3 streams of arguments above which use "utf-8" for writing out the entry. All those other 3 streams of arguments, use "utf-8" for reading too (which is OK). For each of those 3 streams of arguments, could you also add another argument stream which uses non-utf8 charset for reading, like you do here? I think that would give this a bit more coverage, especially for cases like the surrogates in comments and entry names.

Reply:
Thanks, these cases got lost during my parameterization. I've added them back now. 

One such case would probably suffice to verify that the bit flag was interpreted correctly, but it doesn't hurt to have some more variation here.